File: shelf/src/main/java/com/toddway/shelf/rx/ShelfSubjectFactory.java
package com.toddway.shelf.rx;

import com.toddway.shelf.ShelfItem;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.Callable;

import io.reactivex.Observable;

public class ShelfSubjectFactory {

    public static <T> CacheSubject<T> create(final ShelfItem shelfItem, final Class<T> type) {
        return CacheSubject.create(new CacheSubject.Source<T>() {
            @Override
            public Observable<T> read() {
                return Observable.fromCallable(new Callable<T>() {
                    @Override
                    public T call() throws Exception {
                        return shelfItem.get(type);
                    }
                });
            }

            @Override
            public Observable<T> write(final T t) {
                return observeWrite(shelfItem, t);
            }
        });
    }

    public static <T> CacheSubject<List<T>> createForList(final ShelfItem shelfItem, final Class<T[]> type) {
        return CacheSubject.create(new CacheSubject.Source<List<T>>() {
            @Override
            public Observable<List<T>> read() {
                return Observable.fromCallable(new Callable<List<T>>() {
                    @Override
                    public List<T> call() throws Exception {
                        T[] array = shelfItem.get(type);
                        return array == null ? new ArrayList<T>() : Arrays.asList(array);
                    }
                });
            }

            @Override
            public Observable<List<T>> write(final List<T> t) {
                return observeWrite(shelfItem, t);
            }
        });
    }

    private static <T> Observable<T> observeWrite(final ShelfItem shelfItem, final T t) {
        return Observable.fromCallable(new Callable<T>() {
            @Override
            public T call() throws Exception {
                shelfItem.put(t);
                return t;
            }
        });
    }
}

File: shelf/src/main/java/com/toddway/shelf/rx/ShelfSubjectFactory.kt
package com.toddway.shelf.rx

import com.toddway.shelf.ShelfItem
import rx.Observable
import java.util.*

object ShelfSubjectFactory {

    fun <T> create(shelfItem: ShelfItem, type: Class<T>): CacheSubject<T> {
        return CacheSubject.create(object : CacheSubject.Source<T> {
            override fun read(): Observable<T> {
                return Observable.fromCallable { shelfItem.get(type) }
            }

            override fun write(t: T): Observable<T> {
                return observeWrite(shelfItem, t)
            }
        })
    }

    fun <T> createForList(shelfItem: ShelfItem, type: Class<Array<T>>): CacheSubject<List<T>> {
        return CacheSubject.create(object : CacheSubject.Source<List<T>> {
            override fun read(): Observable<List<T>> {
                return Observable.fromCallable {
                    val array = shelfItem.get(type)
                    if (array == null) ArrayList() else Arrays.asList(*array)
                }
            }

            override fun write(t: List<T>): Observable<List<T>> {
                return observeWrite(shelfItem, t)
            }
        })
    }

    private fun <T> observeWrite(shelfItem: ShelfItem, t: T): Observable<T> {
        return Observable.fromCallable {
            shelfItem.put(t)
            t
        }
    }
}
